### PR TITLE
Updated rapid drivers to work with the festo rail

### PR DESCRIPTION
**Pull request by bhomberg**

### CODE REVIEW

**Comment by bhomberg** [comment]
I don't know why github doesn't seem to be noticing the common lines in ROS_common.sys and ROS_socket.sys (a quick sample doesn't suggest whitespace issues....) -- if you have any suggestions for how to fix that, please let me know!  Other than that, this is ready for review.

I can add a readme if you think it would be necessary -- I have an additional ABB "task" and ABB "automatic loading of modules" compared to the ABB ROS Industrial tutorial.

**Comment by bhomberg** [comment]
@jonbinney tagging so you see this is happening / if you have a chance to do a review

**Comment by gavanderhoorn** [comment]
@bhomberg: this is typically something that has to do with line-endings having changed. As that happens to all lines, git (not just github) will show that as a change to the entire file.

**Comment by bhomberg** [comment]
@gavanderhoorn -- ah, makes sense.  Is there a default for this directory that I should default to?  I realize I've been saving in various different editors across the windows machine w/ RobotStudio and my ROS laptop.

**Comment by gavanderhoorn** [comment]
While it is possible to configure a "per directory default" with git, that has not been done in `abb_driver/rapid`. This would seem to be a local configuration issue.

**Comment by gavanderhoorn** [comment]
As a general comment btw: would you guys be interested in making this a little more generic and then upstream it to `ros-industrial/abb`?

**Comment by bhomberg** [comment]
@stwirth -- fixed file endings, ready for review!

@gavanderhoorn -- we're certainly not opposed to making this more generic and pushing it upstream!  (We probably don't have time soon, but would be interested eventually).  I'm not sure how well this generalizes / meshes with preexisting conventions, though.  Right now, we're sending rail/arm trajectories as one combined trajectory (instead of using the external axes).  This PR is also dependent upon a large configuration file integrating the rail into the ABB controller (which we actually had ABB come and install for us) and all of the commands are specific for this rail's interface -- I believe that other external axes would have a maker-specific interface that would need to be configured (replacing ROS_festo.sys).  Did you have a specific idea of how this would be generalized / which components you'd like to see included?

**Comment by gavanderhoorn** [comment]
I've not done this myself, but I could imagine coming up with some sort of API for interfacing with Rapid modules that provide an implementation of an interface to a custom external axis (if something like that doesn't already exist, of course).

Some functions I can come up with:

 - `initialise(..)`
 - `reset(..)`
 - `is_ok(..)`
 - `get_pose(..)`
 - `move_to(..)`
 - `move_to_rel(..)`

etc. Perhaps an internal state machine to properly coordinate all of that and also some conversion functions to go to and from internal representations to whatever the 'generic' pose representation should be (probably metres for linear axes) and the internal mfg representation (probably millimeters).

Once we have that, the driver could call those functions instead of the mfg-specific ones. For your festo module, the shim would call into `ROS_festo.sys`. Alternatively, `ROS_festo.sys` could implement the functions directly.

> Right now, we're sending rail/arm trajectories as one combined trajectory (instead of using the external axes).

You're saying this as if it would be different from how you think it would be done otherwise. I'm not sure I follow. For ROS, the external axis is essentially just one additional joint. `JointTrajectoryPoint`s would encode the values for it in metres. `abb_driver` passes those along untouched.

The Rapid code could should check how many axes are for the primary mech unit (`GetNextMechUnit(.., \NoOfAxes)`?) and then use the rest of the data as external. Alternatively we could make this configurable by the user in some way (as part of the system setup).

File: abb_driver/rapid/ROS_festo.sys [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    PROC Festo_Homing()
+        TPWrite "Home Festo";
+        SetGO CPOS_Start,5;
+    ENDPROC
+
+    PROC Festo_move()
+        SetGO CPOS_Start,1;
+        SETGO Velocity,FESTO_rail_velocity;
+        TPWrite "written festo pos: " \Num:=FESTO_rail_position;
+        IF(FESTO_rail_position > 2500) THEN FESTO_rail_position := 2499; ENDIF
```
Comment (stwirth) [inline abb_driver/rapid/ROS_festo.sys]: make IF clause equal to the below one? 
`IF(FESTO_rail_position > 2500) THEN FESTO_rail_position := 2500`
or
`IF(FESTO_rail_position > 2499) THEN FESTO_rail_position := 2499`

File: abb_driver/rapid/ROS_festo_init.mod [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    IF (NOT has_been_init) THEN
+        ! Initialize Festo
+        Festo_init;
+    
+        ! Home Festo
+        Festo_homing;
+        
+        has_been_init := TRUE;
+        
+        WaitTime 5;
```
Comment (stwirth) [inline abb_driver/rapid/ROS_festo_init.mod:16]: why wait here?

File: abb_driver/rapid/ROS_motion.mod [hunk 3]
```diff
@@ -47,25 +49,33 @@ PROC main()
...
                 stop_mode := DEFAULT_CORNER_DIST;  ! assume we're smoothing between points
-                IF (current_index = trajectory_size) stop_mode := fine;  ! stop at path end
-
+                IF (current_index = trajectory_size) THEN
+                    stop_mode := fine;  ! stop at path end
+                ENDIF
                 ! Execute move command
-                IF (NOT skip_move)
+                IF (NOT skip_move) THEN
+		            FESTO_rail_position := trajectory{current_index}.rail_position;
```
Comment (stwirth) [inline abb_driver/rapid/ROS_motion.mod:71]: indentation looks off here

File: abb_driver/rapid/ROS_motion.mod [hunk 4]
```diff
@@ -47,25 +49,33 @@ PROC main()
...
+                    ! TODO(bhomberg): set velocity to max?
+                    FESTO_rail_velocity := 100;
```
Comment (stwirth) [inline abb_driver/rapid/ROS_motion.mod]: Can you read that somehow from the trajectory (distance / expected duration)?

**Review by stwirth — COMMENTED** [review]
I don't understand the syntax much... Just a few small comments.

File: abb_driver/rapid/ROS_festo.sys (hunk 1 above)
Comment (gavanderhoorn) [reply abb_driver/rapid/ROS_festo.sys]: Suggestion: refactor checking and enforcing joint limits into a separate function. Erroring-out on poses that violate them would also be an option (and probably more consistent with 'native' joints).

File: abb_driver/rapid/ROS_festo_init.mod (hunk 2 above)
Comment (bhomberg) [reply abb_driver/rapid/ROS_festo_init.mod:16]: The wait is for the robot to home itself -- homing takes a while -- the rail moves slowly from wherever it is to one side, where it marks as 0 the place where there's a push against it / current spike.  (For now -- we're adding limit switches to make this more accurate).

File: abb_driver/rapid/ROS_motion.mod (hunk 4 above)
Comment (bhomberg) [reply abb_driver/rapid/ROS_motion.mod]: That's my eventual plan, but it's not particularly straightforward, unfortunately -- we set the velocity between 0% and 100% of the configured max velocity number buried deep in config files.  Ideally we'd access what that value is somehow (and at the very least I need to go look up what it's currently set to).  I can add a TODO, though.

File: abb_driver/rapid/ROS_stateServer.mod [hunk 5]
```diff
@@ -29,7 +29,7 @@ MODULE ROS_stateServer
 ! WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 LOCAL CONST num server_port := 11002;
-LOCAL CONST num update_rate := 0.10;  ! broadcast rate (sec)
+LOCAL CONST num update_rate := 0.01;  ! broadcast rate (sec)
```
Comment (stwirth) [inline abb_driver/rapid/ROS_stateServer.mod:32]: Does that work? Do we get 100Hz feedback now?

Comment (bhomberg) [reply abb_driver/rapid/ROS_stateServer.mod:32]: Not quite, but it's better than 10Hz -- rostopic hz estimate 83-34Hz.

**Comment by stwirth** [comment]
@bhomberg is this ready to merge? (I don't have write access, so I can't do it)

**Comment by bhomberg** [comment]
@stwirth -- yeah, let's merge it.  It's not exactly what we're running now since we're messing with parameters for motion planning, but it's a start and covers the key points of rail integration.

**Pull request merged**